spring-session/src/main/java/org/springframework/session/IndexResolver.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2014-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.session;
+
+/**
+ * Strategy interface for resolving the value of the {@link Session}'s index.
+ *
+ * @param <S> the type of Session being handled
+ * @author Vedran Pavic
+ * @since 1.3.0
+ * @see FindByIndexNameSessionRepository
+ */
+public interface IndexResolver<S extends Session> {
+
+	/**
+	 * Resolve the value of the given index from the session.
+	 * @param session the session
+	 * @param indexName the index name to resolve
+	 * @return the resolved index value, {@code null} if none found
+	 */
+	String resolve(S session, String indexName);
```

Comment:
> This requires another API because the current interface must have knowledge of the index name.

Yes, that's by design since I originally intended ```IndexResolver``` implementations to be single purpose. Each ```FindByIndexNameSessionRepository``` implementation holds the knowledge of its supported indexes anyway, doesn't it? Or did you mean _another API_ as in API targeted at users? Perhaps ```FindByIndexNameSessionRepository``` interface itself should communicate the index names it is capable of resolving?

> Why not? Seems you could do this:

That kind of implementation is exactly what I wanted to avoid :) Your ```GemfireIndexResolver``` will get messy once other ```IndexResolver``` resolvers get in there too since you're relying on the concrete ```IndexResolver``` implementation.

Regarding your proposed ```IndexResolver``` API and Gemfire implementation, consider things from perspective of current usage of ```PrincipalNameResolver```s and equivalent logic - you'd really like to run the resolving for all the indexes and then extract the single value you're interested in instead of current ```PrincipalNameResolver#resolvePrincipal```?